Feat(web-react): Introduce Card component #1535 (#1757)
lmc-eu/spirit-design-system: packages/web-react/src/components/Card/README.md

| ---------------------- | -------------------------------------------- | ------- | -------- | ------------------------------------------- |
| `hasFilledHeightClass` | `bool`                                       | `false` | ✕        | Whether the image fill the height of a Card |
| `isExpanded`           | `bool`                                       | `false` | ✕        | Whether the media have space around         |
| `size`                 | [[Size dictionary][dictionary-size], `auto`] | `auto`  | ✕        | Size of the image media                     |

On top of the API options, the components accept [additional attributes][readme-additional-attributes].
If you need more control over the styling of a component, you can use [style props][readme-style-props]
and [escape hatches][readme-escape-hatches].

## Card Eyebrow

Basic usage:

```jsx
<CardEyebrow>…</CardEyebrow>
```

## Card Title & Card Link

Basic usage:

```jsx
<CardTitle>…</CardTitle>
```

Advanced example:

```jsx
<CardTitle elementType="h2" isHeading={false}>
  <CardLink href="#">…</CardLink>
</CardTitle>
```

> ### API

Reviewer: we need api table for cardlink too, please

| Name          | Type          | Default | Required | Description                                |
| ------------- | ------------- | ------- | -------- | ------------------------------------------ |
| `elementType` | `ElementType` | `h4`    | ✕        | Type of element                            |
| `isHeading`   | `bool`        | `true`  | ✕        | Whether the title is rendered as a heading |

On top of the API options, the components accept [additional attributes][readme-additional-attributes].
If you need more control over the styling of a component, you can use [style props][readme-style-props]
and [escape hatches][readme-escape-hatches].

## Card Body

Basic usage:

```jsx
<CardBody>…</CardBody>
```

> ## Card Actions

Reviewer: Card footer please

# Card

Basic usage:

```jsx
<Card>…</Card>
```

Advanced example:

```jsx
<Card elementType="article" direction="horizontal" isBoxed>
  …
</Card>
```

> Card composition example:

Reviewer 2: After updating according to the `web` README, this will become the "Full Example" section.

| `elementType` | `ElementType`                                                         | `article`  | ✕        | Type of element                                |
| `isBoxed`     | `bool`                                                                | `false`    | ✕        | Whether the Card have border                   |

On top of the API options, the components accept [additional attributes][readme-additional-attributes].
If you need more control over the styling of a component, you can use [style props][readme-style-props]
and [escape hatches][readme-escape-hatches].

## Card Media

Basic usage:

```jsx
<CardMedia>{/* <img /> */}</CardMedia>
```

Advanced example:

```jsx
<CardMedia size="small" isExpanded hasFilledHeightClass>
  {/* <img /> */}
</CardMedia>
```

### API

| Name                   | Type                                         | Default | Required | Description                                 |
| ---------------------- | -------------------------------------------- | ------- | -------- | ------------------------------------------- |

> | `hasFilledHeightClass` | `bool`                                       | `false` | ✕        | Whether the image fill the height of a Card |

Reviewer 2: ```suggestion
| `hasFilledHeight` | `bool`                                       | `false` | ✕        | Whether the image fills the height of the Card |
```

> | `isExpanded`           | `bool`                                       | `false` | ✕        | Whether the media have space around         |

Reviewer 2: ```suggestion
| `isExpanded`           | `bool`                                       | `false` | ✕        | Whether the media has space around         |
```

> | `size`                 | [[Size dictionary][dictionary-size], `auto`] | `auto`  | ✕        | Size of the image media                     |

Reviewer 2: ```suggestion
| `size`                 | [[Size dictionary][dictionary-size], `auto`] | `auto`  | ✕        | Size of the media                     |
```

Not only image, it works for video too! 🙂

On top of the API options, the components accept [additional attributes][readme-additional-attributes].
If you need more control over the styling of a component, you can use [style props][readme-style-props]
and [escape hatches][readme-escape-hatches].

## Card Eyebrow

Basic usage:

```jsx
<CardEyebrow>…</CardEyebrow>
```

## Card Title & Card Link

Basic usage:

```jsx
<CardTitle>…</CardTitle>
```

Advanced example:

```jsx
<CardTitle elementType="h2" isHeading={false}>
  <CardLink href="#">…</CardLink>
</CardTitle>
```

> ### API - Card Title

Reviewer 2: Minor detail: In other components, I've seen it in the form of "XXX API".

| ------------- | ------------- | ------- | -------- | ------------------------------------------ |
| `elementType` | `ElementType` | `h4`    | ✕        | Type of element                            |
| `isHeading`   | `bool`        | `true`  | ✕        | Whether the title is rendered as a heading |

On top of the API options, the components accept [additional attributes][readme-additional-attributes].
If you need more control over the styling of a component, you can use [style props][readme-style-props]
and [escape hatches][readme-escape-hatches].

### API - Card Link

`CardLink` utilizes the `Link` component internally and shares its API. For more details, please refer to the [Link component documentation][link-readme].

## Card Body

Basic usage:

```jsx
<CardBody>…</CardBody>
```

## Card Footer

Basic usage:

```jsx
<CardFooter>…</CardFooter>
```

Advanced example:

```jsx
<CardFooter alignmentX="right">…</CardFooter>
```

### API

| Name         | Type                                          | Default | Required | Description               |
| ------------ | --------------------------------------------- | ------- | -------- | ------------------------- |

> | `alignmentX` | [AlignmentX dictionary][dictionary-alignment] | `left`  | ✕        | Alignment of Footer Items |

Reviewer 2: ```suggestion
| `alignmentX` | [AlignmentX dictionary][dictionary-alignment] | `left`  | ✕        | Alignment of footer content |
```

In the vertical Card layout, the media is always expanded to the full width of the CardBody content. For boxed Cards,
the media can be even expanded [to the edges](#expanding-the-media) of the Card.

- `auto` (default)
- `small`
- `medium`
- `large`

For example:

```jsx
<CardMedia size="small">{/* … */}</CardMedia>
```

ℹ️ The Card automatically prevents the media from overflowing the Card container or even pushing the subsequent
CardBody content out of the Card. In such cases, the media will be cropped to fit the Card container.

### Expanding the Media

To expand the media to the full width or height of a boxed Card, use the `isExpanded` prop. This option is
available for both vertical and horizontal (including reversed horizontal) Card layouts.

```jsx
<Card>
  <CardMedia isExpanded>{/* … */}</CardMedia>
  <CardBody>{/* … */}</CardBody>
</Card>
```

Additionally, there is a `filledHeight` prop that expands the media to match the height of the CardBody
content. This option works with both boxed and non-boxed Card, but is only available in the horizontal Card layout.

```jsx

> <CArd direction="horizontal">

Reviewer 3: **nitpick (if-minor):** 
```suggestion
<Card direction="horizontal">
```

### API

| Name                   | Type                                         | Default | Required | Description                                 |
| ---------------------- | -------------------------------------------- | ------- | -------- | ------------------------------------------- |
| `hasFilledHeightClass` | `bool`                                       | `false` | ✕        | Whether the image fill the height of a Card |
| `isExpanded`           | `bool`                                       | `false` | ✕        | Whether the media has space around          |
| `size`                 | [[Size dictionary][dictionary-size], `auto`] | `auto`  | ✕        | Size of the image media                     |

On top of the API options, the components accept [additional attributes][readme-additional-attributes].
If you need more control over the styling of a component, you can use [style props][readme-style-props]
and [escape hatches][readme-escape-hatches].

## CardLogo

CardLogo is an optional subcomponent that displays a logo. To achieve the best visual result, use the PartnerLogo
subcomponent.

```jsx
<CardLogo>

>   <PartnerLogo class="PartnerLogo">

Reviewer 3: **suggestion (if-minor):** 
```suggestion
  <PartnerLogo>
```

or use `UNSAFE_className`.

```jsx
<CardEyebrow>Content options</CardEyebrow>
<CardTitle>Card Title</CardTitle>
```

## CardFooter

Use CardFooter for actions or any other content at the bottom of the Card. When using Cards with CardFooter in a Grid,
the CardFooters will automatically line up.

```jsx
<CardFooter>{/* … */}</CardFooter>
```

### Footer Alignment

The footer can be horizontally aligned to the start, center, or end of the Card. To align the footer, use one of the
following `alignmentX` prop values:

- `left` (default)
- `center`
- `right`

### API

| Name         | Type                                          | Default | Required | Description                 |
| ------------ | --------------------------------------------- | ------- | -------- | --------------------------- |
| `alignmentX` | [AlignmentX dictionary][dictionary-alignment] | `left`  | ✕        | Alignment of footer content |

On top of the API options, the components accept [additional attributes][readme-additional-attributes].
If you need more control over the styling of a component, you can use [style props][readme-style-props]
and [escape hatches][readme-escape-hatches].

> ## Full Example

Reviewer: please move example to the end of the file

Reviewer: maybe generally, apply all changes done in the web package

When you put it all together:

```jsx
<Card isBoxed>
  <CardMedia>
    <img src="…" alt="" />
  </CardMedia>
  <CardLogo>
    <PartnerLogo>
      <img src="…" alt="Logo" />
    </PartnerLogo>
  </CardLogo>
  <CardBody>
    <CardEyebrow>Content options</CardEyebrow>
    <CardTitle>
      <CardLink href="#">Card Title</CardLink>
    </CardTitle>
    <p>Card content</p>
  </CardBody>
  <CardFooter>
    <ButtonLink>Primary</ButtonLink>
    <ButtonLink>Secondary</ButtonLink>
  </CardFooter>
</Card>
```

## Card Grid

In a typical use case, you will display multiple Cards in a [Grid][grid].

```jsx
<Grid cols={{ mobile: 1, tablet: 2, desktop: 3 }}>
  <Card>{/* … */}</Card>
  <Card>{/* … */}</Card>
  <Card>{/* … */}</Card>
</Grid>
```

Depending on your situation, you may want to use the list semantics. And it will work!

```jsx

> <Grid cols={{ mobile: 1, tablet: 2, desktop: 3 }}>

Reviewer: elementtype ul pls

ℹ️ Don't worry, any interactive elements inside the Card (like links or buttons) will still work as expected.

ℹ️ The text content of CardBody remains selectable and copyable even when the whole Card is clickable.

### API

| Name          | Type          | Default | Required | Description     |
| ------------- | ------------- | ------- | -------- | --------------- |
| `elementType` | `ElementType` | `a`     | ✕        | Type of element |

On top of the API options, the components accept [additional attributes][readme-additional-attributes].
If you need more control over the styling of a component, you can use [style props][readme-style-props]
and [escape hatches][readme-escape-hatches].

### Linking the Media

In most cases, using just a single link in the CardTitle and
[making the whole card clickable](#making-the-whole-card-clickable) is the best approach in terms of accessibility.
The Card will have a single accessible link which will be announced by screen readers.

However, if you cannot use the CardLink subcomponent, and you still need to make the media clickable, you can wrap the
CardMedia image in a link:

```jsx
<CardMedia>

>   <a href="#" aria-hidden="true">

Reviewer 3: **nitpick (if-minor):** Can you please use a `Link` component?

    <img src="https://via.placeholder.com/300x200" alt="" />
  </a>
</CardMedia>
```

👉 Please note that the `aria-hidden="true"` attribute is used to hide the link from screen readers so the user is not
confused by too many links in the Card.

### The “Read More” Use Case

For article previews or similar use cases, you may want to display a limited amount of text content with a “Read More”
link. For optimum accessibility, you should only provide this in form of a text node, not a button or a link:

```jsx
<CardTitle>
  <CardLink href="#">Card title</a>
</CardTitle>
<p>{/* … */}</p>

> <!-- DON'T DO THIS -->

Reviewer: we can use jsx comments

> <a href="#" class="link-primary link-underlined">Read more</a>

Reviewer 3: **question (if-minor):** A Link component?

<!-- This is correct -->
<div class="link-primary link-underlined">Read more</div>
```

This way, the Card will only have a single accessible link which will be announced by screen readers.

ℹ️ A big shout-out to [Ondřej Pohl][ondrej-pohl] for sharing many of these best practices!

> [grid]: https://github.com/lmc-eu/spirit-design-system/blob/main/packages/web/src/scss/components/Grid/README.md

Reviewer: please link react package
